streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/CombinedKey.java

```
@@ -38,8 +38,22 @@ public KP getPrimaryKey() {
         return primaryKey;
     }
 
-    public boolean equals(final KF foreignKey, final KP primaryKey) {
-        return this.foreignKey.equals(foreignKey) && this.primaryKey.equals(primaryKey);
+    @Override
+    public int hashCode() {
+        return Objects.hash(foreignKey, primaryKey);
+    }
```

Do we need to implement hashCode? It's not free of side effects. If we don't actually need it, but also want to have a correct hashCode/equals contract, you can throw an UnsupportedOperationException from the hashCode method, effectively preventing anyone from using these objects in hash collections.

Why would you want to do that? Mostly if you can't guarantee that all fields are immutable. OTOH, if all fields are immutable, then it's safe to just implement it.

---

Both fields of `CombinedKey` are immutable:

https://github.com/apache/kafka/blob/2ff4c0a364379d231edfe1836042574359b1023e/streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/CombinedKey.java#L22-L24

---

Sounds good to me.